File: frontend/src/types/production.ts
// Production module types

export type ProdBatchStatus = 'open' | 'closed';
export type ProdShift = 'Morning' | 'Night';
export type ProdPackType = '4KG_SEMI' | '1KG_FG' | '0.5KG_FG' | 'BULK_KG';
export type ProdProductType = 'forming' | 'hot_process';
export type PackApplicableType = 'forming' | 'hot_process' | 'both';

// ----- Pack Type Config -----
export interface PackTypeConfig {
  id: number;
  code: string;
  name: string;
  applicable_type: PackApplicableType;
  nominal_weight_kg: number | null;
  is_active: boolean;
  created_at: string;
}

export interface PackTypeConfigCreate {
  code: string;
  name: string;
  applicable_type?: PackApplicableType;
  nominal_weight_kg?: number | null;
}

export interface PackTypeConfigUpdate {
  name?: string;
  applicable_type?: PackApplicableType;
  nominal_weight_kg?: number | null;
  is_active?: boolean;
}

export interface FormingOption {
  id: number;
  code: string;
  name: string;
  product_type: ProdProductType;
  ccp_limit_temp: string | null;
  pack_size_kg: number | null;
  loss_rate_warn_pct: number | null;
}

// ----- Products -----
export interface ProdProduct {
  id: number;
  code: string;
  name: string;
  ccp_limit_temp: string;
  pack_size_kg: number | null;
  loss_rate_warn_pct: number | null;
  product_type: ProdProductType;
  inv_item_id: number | null;
  is_active: boolean;
  created_at: string;
}

export interface ProdProductCreate {
  code: string;
  name: string;
  ccp_limit_temp?: string;
  pack_size_kg?: number | null;
  loss_rate_warn_pct?: number | null;
  product_type?: ProdProductType;
  inv_item_id?: number | null;
}

export interface ProdProductUpdate {
  name?: string;
  ccp_limit_temp?: string;
  pack_size_kg?: number | null;
  loss_rate_warn_pct?: number | null;
  product_type?: ProdProductType;
  inv_item_id?: number | null;
  is_active?: boolean;
}

// ----- Forming Trolleys -----
export interface ProdFormingTrolley {
  id: number;
  batch_id: number;
  trolley_no: string;
  sampled_tray_count: number;
  sampled_gross_weight_sum_kg: number;
  tray_tare_weight_kg: number;
  total_trays_on_trolley: number;
  partial_trays_count: number;
  partial_fill_ratio: number;
  avg_tray_net_weight_kg: number | null;
  equivalent_tray_count: number | null;
  estimated_net_weight_kg: number | null;
  remark: string | null;
}

export interface ProdFormingTrolleyCreate {
  trolley_no: string;
  sampled_tray_count: number;
  sampled_gross_weight_sum_kg: number;
  tray_tare_weight_kg: number;
  total_trays_on_trolley: number;
  partial_trays_count?: number;
  partial_fill_ratio?: number;
  remark?: string;
}

// ----- Packing -----
export interface ProdPackingRecord {
  id: number;
  batch_id: number;
  pack_type: ProdPackType;
  product_id: number | null;
  product_name: string | null;
  inv_item_id: number | null;
  inv_item_name: string | null;
  bag_count: number;
  nominal_weight_kg: number;
  theoretical_total_weight_kg: number | null;
  remark: string | null;
}

export interface ProdPackingRecordCreate {
  pack_type: ProdPackType;
  product_id?: number | null;
  inv_item_id?: number;
  bag_count: number;
  nominal_weight_kg: number;
  remark?: string;
}

export interface ProdPackingTrim {
  id: number;
  batch_id: number;
  trim_type: string;
  weight_kg: number;
  remark: string | null;
}

export interface ProdPackingTrimCreate {
  trim_type: string;
  weight_kg: number;
  remark?: string;
}

export interface ProdPackingSaveRequest {
  records: ProdPackingRecordCreate[];
  trims: ProdPackingTrimCreate[];
}

// ----- Hot Inputs (熱加工投料) -----
export interface ProdHotInput {
  id: number;
  prod_batch_id: number;
  seq: number;
  weight_kg: string;
  notes: string | null;
  created_at: string;
}

export interface ProdHotInputCreate {
  weight_kg: number | string;
  notes?: string;
}

// ----- Batches -----
export interface ProdBatch {
  id: number;
  batch_code: string;
  product_code: string;
  product_name: string;
  production_date: string;
  shift: ProdShift | null;
  spec_piece_weight_g: number;
  start_time: string | null;
  end_time: string | null;
  status: ProdBatchStatus;
  operator: string | null;
  supervisor: string | null;
  estimated_forming_net_weight_kg: string | null;
  estimated_forming_pieces: number | null;
  input_weight_kg: string | null;
  inv_stock_doc_id: number | null;
  created_at: string;
  trolleys: ProdFormingTrolley[];
  packing_records: ProdPackingRecord[];
  packing_trims: ProdPackingTrim[];
  hot_inputs: ProdHotInput[];
}

export interface ProdBatchCreate {
  product_code: string;
  product_name: string;
  production_date: string;
  shift?: ProdShift;
  spec_piece_weight_g?: number;
  start_time?: string;
  operator?: string;
  supervisor?: string;
  input_weight_kg?: number;
}

export interface ProdBatchUpdate {
  shift?: ProdShift;
  spec_piece_weight_g?: number;
  start_time?: string;
  end_time?: string;
  operator?: string;
  supervisor?: string;
  input_weight_kg?: number;
}

// ----- Hot Process Balance -----
export interface HotProcessBalance {
  input_weight_kg: string | null;
  packed_weight_kg: string;
  loss_weight_kg: string | null;
  loss_rate: string | null;
}

// ----- Repack -----
export interface ProdRepackInput {
  id: number;
  repack_job_id: number;
  from_batch_id: number | null;
  from_batch_code: string | null;
  product_id: number | null;
  product_name: string | null;
  bag_count: number;
  nominal_weight_kg: number;
  total_weight_kg: number | null;
}

export interface ProdRepackInputCreate {
  from_batch_id?: number;
  product_id?: number;
  bag_count: number;
  nominal_weight_kg: number;
}

export interface ProdRepackOutput {
  id: number;
  repack_job_id: number;
  pack_type: ProdPackType;
  product_id: number | null;
  product_name: string | null;
  bag_count: number;
  nominal_weight_kg: number;
  total_weight_kg: number | null;
}

export interface ProdRepackOutputCreate {
  pack_type: ProdPackType;
  product_id?: number;
  bag_count: number;
  nominal_weight_kg: number;
}

export interface ProdRepackTrim {
  id: number;
  repack_job_id: number;
  trim_type: string;
  weight_kg: number;
  remark: string | null;
}

export interface ProdRepackTrimCreate {
  trim_type: string;
  weight_kg: number;
  remark?: string;
}

export interface ProdRepackJob {
  id: number;
  new_batch_code: string;
  date: string;
  operator: string | null;
  remark: string | null;
  created_at: string;
  inputs: ProdRepackInput[];
  outputs: ProdRepackOutput[];
  trims: ProdRepackTrim[];
}

export interface ProdRepackJobCreate {
  date: string;
  operator?: string;
  remark?: string;
}

// ----- Calculation responses -----
export interface FormingTotals {
  total_net_weight_kg: number;
  total_pieces: number;
  duration_minutes: number | null;
  pieces_per_hour: number | null;
}

export interface PackingTotals {
  forming_input_kg: number;
  total_4kg_kg: number;
  total_retail_kg: number;
  total_trim_kg: number;
  output_total_kg: number;
  loss_kg: number;
  loss_rate: number;
}

export interface RepackTotals {
  input_total_kg: number;
  output_total_kg: number;
  trim_total_kg: number;
  loss_kg: number;
  loss_rate: number;
}

File: backend/app/schemas/production.py
"""Production module schemas (生產系統)."""

from datetime import date, datetime
from decimal import Decimal
from typing import Optional, List

from pydantic import BaseModel, ConfigDict, Field


# ---------------------------------------------------------------------------
# Pack type config (包裝類型設定)
# ---------------------------------------------------------------------------

class ProdPackTypeConfigCreate(BaseModel):
    code: str = Field(..., max_length=50)
    name: str = Field(..., max_length=200)
    applicable_type: str = "both"   # forming | hot_process | both
    nominal_weight_kg: Optional[Decimal] = None


class ProdPackTypeConfigUpdate(BaseModel):
    name: Optional[str] = Field(None, max_length=200)
    applicable_type: Optional[str] = None
    nominal_weight_kg: Optional[Decimal] = None
    is_active: Optional[bool] = None


class ProdPackTypeConfigResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    code: str
    name: str
    applicable_type: str
    nominal_weight_kg: Optional[Decimal] = None
    is_active: bool
    created_at: datetime


# ---------------------------------------------------------------------------
# Product (產品)
# ---------------------------------------------------------------------------

class ProdProductCreate(BaseModel):
    code: str = Field(..., max_length=50)
    name: str = Field(..., max_length=200)
    ccp_limit_temp: Decimal = Decimal("75.00")
    pack_size_kg: Optional[Decimal] = None
    loss_rate_warn_pct: Optional[Decimal] = None
    product_type: str = "forming"
    inv_item_id: Optional[int] = None


class ProdProductUpdate(BaseModel):
    name: Optional[str] = Field(None, max_length=200)
    ccp_limit_temp: Optional[Decimal] = None
    pack_size_kg: Optional[Decimal] = None
    loss_rate_warn_pct: Optional[Decimal] = None
    product_type: Optional[str] = None
    inv_item_id: Optional[int] = None
    is_active: Optional[bool] = None


class ProdProductResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    code: str
    name: str
    ccp_limit_temp: Decimal
    pack_size_kg: Optional[Decimal] = None
    loss_rate_warn_pct: Optional[Decimal] = None
    product_type: str = "forming"
    inv_item_id: Optional[int] = None
    is_active: bool
    created_at: datetime


class FormingOption(BaseModel):
    id: int
    code: str
    name: str
    product_type: str
    ccp_limit_temp: Optional[Decimal] = None
    pack_size_kg: Optional[Decimal] = None
    loss_rate_warn_pct: Optional[Decimal] = None


# ---------------------------------------------------------------------------
# Forming trolley (成型台車)
# ---------------------------------------------------------------------------

class ProdFormingTrolleyCreate(BaseModel):
    trolley_no: str = Field(..., max_length=20)
    sampled_tray_count: int
    sampled_gross_weight_sum_kg: Decimal
    tray_tare_weight_kg: Decimal
    total_trays_on_trolley: int
    partial_trays_count: int = 0
    partial_fill_ratio: Decimal = Decimal("0.5")
    remark: Optional[str] = None


class ProdFormingTrolleyResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    batch_id: int
    trolley_no: str
    sampled_tray_count: int
    sampled_gross_weight_sum_kg: Decimal
    tray_tare_weight_kg: Decimal
    total_trays_on_trolley: int
    partial_trays_count: int
    partial_fill_ratio: Decimal
    avg_tray_net_weight_kg: Optional[Decimal] = None
    equivalent_tray_count: Optional[Decimal] = None
    estimated_net_weight_kg: Optional[Decimal] = None
    remark: Optional[str] = None


# ---------------------------------------------------------------------------
# Packing record (包裝記錄)
# ---------------------------------------------------------------------------

class ProdPackingRecordCreate(BaseModel):
    pack_type: str = Field(..., max_length=20)
    product_id: Optional[int] = None
    inv_item_id: Optional[int] = None
    bag_count: int
    nominal_weight_kg: Decimal
    remark: Optional[str] = None


class ProdPackingRecordResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    batch_id: int
    pack_type: str
    product_id: Optional[int] = None
    product_name: Optional[str] = None
    inv_item_id: Optional[int] = None
    inv_item_name: Optional[str] = None
    bag_count: int
    nominal_weight_kg: Decimal
    theoretical_total_weight_kg: Optional[Decimal] = None
    remark: Optional[str] = None


# ---------------------------------------------------------------------------
# Packing trim (包裝邊角料)
# ---------------------------------------------------------------------------

class ProdPackingTrimCreate(BaseModel):
    trim_type: str = Field(..., max_length=100)
    weight_kg: Decimal
    remark: Optional[str] = None


class ProdPackingTrimResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    batch_id: int
    trim_type: str
    weight_kg: Decimal
    remark: Optional[str] = None


# ---------------------------------------------------------------------------
# Packing save (bulk save request)
# ---------------------------------------------------------------------------

class ProdPackingSaveRequest(BaseModel):
    records: List[ProdPackingRecordCreate]
    trims: List[ProdPackingTrimCreate]


# ---------------------------------------------------------------------------
# Batch (批次)
# ---------------------------------------------------------------------------

class ProdBatchCreate(BaseModel):
    product_code: str = Field(..., max_length=50)
    product_name: str = Field(..., max_length=200)
    production_date: date
    shift: Optional[str] = None
    spec_piece_weight_g: Decimal = Decimal("0")
    start_time: Optional[datetime] = None
    operator: Optional[str] = Field(None, max_length=100)
    supervisor: Optional[str] = Field(None, max_length=100)
    input_weight_kg: Optional[Decimal] = None


class ProdBatchUpdate(BaseModel):
    shift: Optional[str] = None
    spec_piece_weight_g: Optional[Decimal] = None
    start_time: Optional[datetime] = None
    end_time: Optional[datetime] = None
    operator: Optional[str] = Field(None, max_length=100)
    supervisor: Optional[str] = Field(None, max_length=100)
    input_weight_kg: Optional[Decimal] = None


class ProdHotInputCreate(BaseModel):
    weight_kg: Decimal = Field(..., gt=0)
    notes: Optional[str] = None


class ProdHotInputResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    prod_batch_id: int
    seq: int
    weight_kg: Decimal
    notes: Optional[str] = None
    created_at: datetime


class ProdBatchResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    batch_code: str
    product_code: str
    product_name: str
    production_date: date
    shift: Optional[str] = None
    spec_piece_weight_g: Decimal
    start_time: Optional[datetime] = None
    end_time: Optional[datetime] = None
    status: str
    operator: Optional[str] = None
    supervisor: Optional[str] = None
    estimated_forming_net_weight_kg: Optional[Decimal] = None
    estimated_forming_pieces: Optional[int] = None
    input_weight_kg: Optional[Decimal] = None
    inv_stock_doc_id: Optional[int] = None
    created_at: datetime
    trolleys: List[ProdFormingTrolleyResponse] = []
    packing_records: List[ProdPackingRecordResponse] = []
    packing_trims: List[ProdPackingTrimResponse] = []
    hot_inputs: List[ProdHotInputResponse] = []


class HotProcessBalanceResponse(BaseModel):
    input_weight_kg: Optional[Decimal] = None
    packed_weight_kg: Decimal
    loss_weight_kg: Optional[Decimal] = None
    loss_rate: Optional[Decimal] = None


class EnterStockRequest(BaseModel):
    location_id: int


# ---------------------------------------------------------------------------
# Repack input (改包投入)
# ---------------------------------------------------------------------------

class ProdRepackInputCreate(BaseModel):
    from_batch_id: Optional[int] = None
    product_id: Optional[int] = None
    bag_count: int
    nominal_weight_kg: Decimal


class ProdRepackInputResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    repack_job_id: int
    from_batch_id: Optional[int] = None
    from_batch_code: Optional[str] = None
    product_id: Optional[int] = None
    product_name: Optional[str] = None
    bag_count: int
    nominal_weight_kg: Decimal
    total_weight_kg: Optional[Decimal] = None


# ---------------------------------------------------------------------------
# Repack output (改包產出)
# ---------------------------------------------------------------------------

class ProdRepackOutputCreate(BaseModel):
    pack_type: str = Field(..., max_length=20)
    product_id: Optional[int] = None
    bag_count: int
    nominal_weight_kg: Decimal


class ProdRepackOutputResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    repack_job_id: int
    pack_type: str
    product_id: Optional[int] = None
    product_name: Optional[str] = None
    bag_count: int
    nominal_weight_kg: Decimal
    total_weight_kg: Optional[Decimal] = None


# ---------------------------------------------------------------------------
# Repack trim (改包邊角料)
# ---------------------------------------------------------------------------

class ProdRepackTrimCreate(BaseModel):
    trim_type: str = Field(..., max_length=100)
    weight_kg: Decimal
    remark: Optional[str] = None


class ProdRepackTrimResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    repack_job_id: int
    trim_type: str
    weight_kg: Decimal
    remark: Optional[str] = None


# ---------------------------------------------------------------------------
# Repack job (改包工單)
# ---------------------------------------------------------------------------

class ProdRepackJobCreate(BaseModel):
    date: date
    operator: Optional[str] = Field(None, max_length=100)
    remark: Optional[str] = None


class ProdRepackJobResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    new_batch_code: str
    date: date
    operator: Optional[str] = None
    remark: Optional[str] = None
    created_at: datetime
    inputs: List[ProdRepackInputResponse] = []
    outputs: List[ProdRepackOutputResponse] = []
    trims: List[ProdRepackTrimResponse] = []


# ---------------------------------------------------------------------------
# Calculation responses (計算結果)
# ---------------------------------------------------------------------------

class FormingTotalsResponse(BaseModel):
    total_net_weight_kg: Decimal
    total_pieces: int
    duration_minutes: Optional[float] = None
    pieces_per_hour: Optional[float] = None


class PackingTotalsResponse(BaseModel):
    forming_input_kg: Decimal
    total_packed_kg: Decimal
    total_trim_kg: Decimal
    output_total_kg: Decimal
    loss_kg: Decimal
    loss_rate: float
    by_pack_type: dict[str, Decimal] = {}


class RepackTotalsResponse(BaseModel):
    input_total_kg: Decimal
    output_total_kg: Decimal
    trim_total_kg: Decimal
    loss_kg: Decimal
    loss_rate: float
